oauth2/oauth2-client/src/main/java/org/springframework/security/oauth2/client/web/OAuth2AuthorizationExchangeUtils.java

```diff
@@ -0,0 +1,93 @@
+/*
+ * Copyright 2002-2019 the original author or authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.springframework.security.oauth2.client.web;
+
+import org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationRequest;
+import org.springframework.security.oauth2.core.endpoint.OAuth2AuthorizationResponse;
+import org.springframework.security.web.util.UrlUtils;
+import org.springframework.util.MultiValueMap;
+import org.springframework.web.util.UriComponentsBuilder;
+
+import java.net.URI;
+import java.util.List;
+import java.util.Map;
+
+/**
+ * Utility methods for OAuth 2.0 Authorization Exchange(Request and Response).
+ *
+ * @author Tadaya Tsuyukubo
+ * @since 5.2
+ * @see OAuth2AuthorizationRequest
+ * @see OAuth2AuthorizationResponse
+ */
+public final class OAuth2AuthorizationExchangeUtils {
```

Comment:
Can you please remove this and keep the changes contained in `OAuth2AuthorizationResponseUtils` and `OAuth2AuthorizationCodeGrantFilter`, as well as the Reactive counterparts. I would like to get this into a patch release.

NOTE: In `processAuthorizationResponse()`, I believe you can build the `redirect-uri` for `OAuth2AuthorizationResponse` as follows:

```java
		MultiValueMap<String, String> params = OAuth2AuthorizationResponseUtils.toMultiMap(request.getParameterMap());
		String redirectUri = UriComponentsBuilder.fromHttpUrl(UrlUtils.buildFullRequestUrl(request))
				.replaceQueryParam(OAuth2ParameterNames.CODE, (Object) null)
				.replaceQueryParam(OAuth2ParameterNames.STATE, (Object) null)
				.build()
				.toUriString();
		OAuth2AuthorizationResponse authorizationResponse = OAuth2AuthorizationResponseUtils.convert(params, redirectUri);
```

Reply:
The `OAuth2AuthorizationResponseUtils` is package scoped to `org.springframework.security.oauth2.client.web`, so cannot reference from provider/validator(`org.springframework.security.oauth2.client.authentication`).

So, ok to make it public??

Reply:
`OAuth2AuthorizationExchangeValidator` does not need to change. What I was suggesting previously is to add the following to `OAuth2AuthorizationResponseUtils`:

```java
	static OAuth2AuthorizationResponse convert(HttpServletRequest request) {
		MultiValueMap<String, String> params = toMultiMap(request.getParameterMap());
		MultiValueMap<String, String> redirectParams = new LinkedMultiValueMap<>(params);
		redirectParams.remove(OAuth2ParameterNames.CODE);
		redirectParams.remove(OAuth2ParameterNames.STATE);
		String redirectUri = UriComponentsBuilder.fromHttpUrl(UrlUtils.buildFullRequestUrl(request))
				.replaceQueryParams(redirectParams)
				.build()
				.toUriString();
		return convert(params, redirectUri);
	}
```

This utility method can than be used in `shouldProcessAuthorizationResponse` and `processAuthorizationResponse`. This will contain the changes only in `OAuth2AuthorizationCodeGrantFilter` and `OAuth2AuthorizationCodeGrantWebFilter`

Reply:
ok, but I think there is be a problem when multiple parameters are presented.

For example, following should match:
- `http://localhost/auth?foo=FOO&bar=BAR`
- `http://localhost/auth?bar=BAR&foo=FOO`

If it does simple string match of `redirect_uri` in `OAuth2AuthorizationExchangeValidator`, they are considered as different.

Reply:
I'm not sure if this is an issue. I'm expecting the provider to perform the redirect using the registered `redirect-uri` **as-is** so it should be the same order as specified in `ClientRegistration.redirect-uri`.

Reply:
I just tested with Okta and it's performing an exact match.

The registered `redirect-uri` is `http://locahost:8080/authorized/okta2?param1=value1&param2=value2`.

If I trigger the Authorization Request using `http://locahost:8080/authorized/okta2?param2=value2&param1=value1` it fails.